components/brave_federated/learning_service.cc

```diff
@@ -0,0 +1,216 @@
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_federated/learning_service.h"
+
+#include <utility>
+#include <vector>
+
+#include "base/logging.h"
+#include "base/task/thread_pool.h"
+#include "base/time/time.h"
+#include "base/timer/timer.h"
+#include "brave/components/brave_federated/communication_adapter.h"
+#include "brave/components/brave_federated/eligibility_service.h"
+#include "brave/components/brave_federated/features.h"
+#include "brave/components/brave_federated/task/federated_task_runner.h"
+#include "brave/components/brave_federated/task/model.h"
+#include "brave/components/brave_federated/task/typing.h"
+#include "brave/components/brave_federated/util/synthetic_dataset.h"
+#include "content/public/browser/browser_task_traits.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "services/network/public/cpp/simple_url_loader.h"
+#include "third_party/abseil-cpp/absl/types/optional.h"
+
+namespace brave_federated {
+
+namespace {
+
+absl::optional<TaskResult> LoadDatasetAndRunTask(
+    std::unique_ptr<FederatedTaskRunner> task_runner) {
+  auto synthetic_dataset = std::make_unique<SyntheticDataset>(500);
+  SyntheticDataset test_dataset = synthetic_dataset->SeparateTestData(50);
+
+  task_runner->SetTrainingData(synthetic_dataset->GetDataPoints());
+  task_runner->SetTestData(test_dataset.GetDataPoints());
+  VLOG(2) << "Model and data set. Task runner initialized.";
+
+  return task_runner->Run();
+}
+
+}  // namespace
+
+LearningService::LearningService(
+    EligibilityService* eligibility_service,
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory)
+    : url_loader_factory_(url_loader_factory),
+      eligibility_service_(eligibility_service),
+      communication_adapter_(
+          std::make_unique<CommunicationAdapter>(url_loader_factory_)) {
+  DCHECK(!init_task_timer_);
+  post_results_policy_ = std::make_unique<const net::BackoffEntry::Policy>(
+      /*.num_errors_to_ignore = */ 0,
+      /*.initial_delay_ms = */
+      features::GetFederatedLearningUpdateCycleInSeconds() * 1000,
+      /*.multiply_factor =*/2.0,
+      /*.jitter_factor =*/0.0,
+      /*.maximum_backoff_ms =*/16 *
+          features::GetFederatedLearningUpdateCycleInSeconds() * 1000,
+      /*.always_use_initial_delay =*/true);
+  post_results_backoff_entry_ =
+      std::make_unique<net::BackoffEntry>(post_results_policy_.get());
+
+  const int init_federated_service_wait_time_in_seconds =
+      brave_federated::features::GetInitFederatedServiceWaitTimeInSeconds();
+  init_task_timer_ = std::make_unique<base::OneShotTimer>();
+  init_task_timer_->Start(
+      FROM_HERE, base::Seconds(init_federated_service_wait_time_in_seconds),
+      this, &LearningService::Init);
+}
+
+LearningService::~LearningService() {
+  if (participating_) {
+    StopParticipating();
+  }
+  if (initialized_) {
+    eligibility_service_->RemoveObserver(this);
```

Comment:
Better to use `base/scoped_observation.h` instead of `Add/RemoveObserver` and `initialized_` guard.